csrc/abstract_tensor.h

```diff
@@ -253,70 +580,262 @@ struct AbstractTensor {
     domain.push_back(std::move(id));
+    info.resize(domain.size());
+    return *this;
+  }
+
+  AbstractTensorWithInfo<Info>& pushBack(AbstractId id, const Info& info) {
+    domain.push_back(std::move(id));
+    info.push_back(info);
     return *this;
   }
 
   template <typename... Args>
-  AbstractTensor& emplaceBack(Args&&... args) {
+  AbstractTensorWithInfo<Info>& emplaceBack(Args&&... args) {
     domain.emplace_back(std::forward<Args>(args)...);
     return *this;
   }
 
   template <typename T>
   bool operator==(T&& t) const {
-    if constexpr (std::is_same_v<AbstractTensor, std::decay_t<T>>) {
-      return domain == t.domain;
-    } else {
-      return domain == std::forward<T>(t);
+    if constexpr (std::is_same_v<
+                      AbstractTensorWithInfo<Info>,
+                      std::decay_t<T>>) {
+      return domain == t.domain && info == t.info;
     }
+    return false;
   }
 
   template <typename T>
   bool operator!=(T&& t) const {
     return !operator==(std::forward<T>(t));
   }
 
-  AbstractTensor& parallelize(int64_t axis, ParallelType parallel_type);
+  AbstractTensorWithInfo<Info>& parallelize(
+      int64_t axis,
+      ParallelType parallel_type) {
+    axis = wrapDim(axis, (int64_t)domain.size());
+    AbstractId::dispatch(DispatchParallelize{}, parallel_type, domain[axis]);
+    return *this;
+  }
+
+  AbstractTensorWithInfo<Info>& split(
+      int64_t axis,
+      Val* factor,
+      bool inner_split = true) {
+    axis = wrapDim(axis, (int64_t)domain.size());
+    auto [outer, inner] = AbstractId::dispatch(
+        DispatchSplit{}, domain[axis], factor, inner_split);
+    std::swap(domain[axis], inner);
+    domain.insert(domain.begin() + axis, outer);
+
+    info.insert(info.begin() + axis, info[axis]);
+
+    return *this;
+  }
+
+  AbstractTensorWithInfo<Info>& split(
+      int64_t axis,
+      int64_t factor,
+      bool inner_split = true) {
+    return split(
+        axis, IrBuilder::create<Val>(factor, DataType::Index), inner_split);
+  }
+
+  AbstractTensorWithInfo<Info>& merge(int64_t axis_o, int64_t axis_i) {
+    axis_o = wrapDim(axis_o, (int64_t)domain.size());
+    axis_i = wrapDim(axis_i, (int64_t)domain.size());
+
+    auto output =
+        AbstractId::dispatch(DispatchMerge{}, domain[axis_o], domain[axis_i]);
+    // axis_o is the outer input of this merge but does not
+    // automatically mean it's an outer domain in this AbstractTensorWithInfo.
+    auto domain_outer_pos = axis_o < axis_i ? axis_o : axis_i;
+    auto domain_inner_pos = axis_o < axis_i ? axis_i : axis_o;
+
+    domain.erase(domain.begin() + domain_inner_pos);
+    std::swap(domain[domain_outer_pos], output);
 
-  AbstractTensor& split(int64_t axis, Val* factor, bool inner_split = true);
-  AbstractTensor& split(int64_t axis, int64_t factor, bool inner_split = true);
+    info[domain_outer_pos] =
+        Info::merge(info[domain_outer_pos], info[domain_inner_pos]);
+    info.resize(info.size() - 1);
```

Comment:
What happens with this resize? Does it drop the last item (which is not what we wanted)?

Reply:
Ah of course. Switched to `erase` instead and added a test. Thanks for noticing!